### ISSUE TITLE
[leo_storage] Better default for watchdog.disk.threshold_disk_util

### ISSUE THREAD

**Issue by mocchira** [leo-project/leofs#693]
In order to prevent leo_storage from being restarted by heart too many times as reported on https://github.com/leo-project/leofs/issues/644#issuecomment-290717176.

One possible solution to this sort of restarts is to set watchdog.disk.threshold_disk_util to more conservative value (the current default = 90: https://github.com/leo-project/leofs/blob/1.3.3-rc3/apps/leo_storage/priv/leo_storage.conf#L143-L144).

We will decide the new default based on as many test cases and as much alike environments with production as possible.

**Comment by yosukehara** [comment]
@windkit 
Currently, the default value of `watchdog.disk.threshold_disk_util` is 90%. We need to recognize its suitable value through some tests using SSD and HDD on our environment. After benchmarking LeoFS 1.3.3-dev, share the report on [leo-project/notes](https://github.com/leo-project/notes), then let's discuss its value.

**Comment by windkit** [comment]
Here is a check of performance with disk watchdog at different disk utilization threshold with Image f4m Write Test Case

### No Disk Watchdog
![grafana_dw_no](https://cloud.githubusercontent.com/assets/360884/24937195/e9da1f6e-1f69-11e7-8434-2d379d5e1bb1.png)
![dw_no](https://cloud.githubusercontent.com/assets/360884/24937196/eb21171a-1f69-11e7-9882-06d99b5c95db.png)

### Disk Watchdog (90%)
![grafana_dw_90](https://cloud.githubusercontent.com/assets/360884/24937198/ef6a9800-1f69-11e7-9eba-1212dc7c3e1a.png)
![dw_90](https://cloud.githubusercontent.com/assets/360884/24937197/ef50fc88-1f69-11e7-8dea-7dc961b6b86e.png)

### Disk Watchdog (50%)
![dw_50](https://cloud.githubusercontent.com/assets/360884/24937202/f4e09622-1f69-11e7-904a-460e1783501d.png)
![grafana_dw_50](https://cloud.githubusercontent.com/assets/360884/24937201/f4e0872c-1f69-11e7-8d6a-34d8953c206f.png)

Given that fact the disk writes are buffered, utilization comes in as spikes, enabling disk utilization watchdog imposes a big write performance hit and the threshold level does not matter much

**Comment by mocchira** [comment]
@windkit 

The load coming from ONLY leo_gateway(s) might not be sufficient to decide what is the best default for threshold_disk_util as leo_watchdog try to not only limit incoming requests but also tweak the rate of consuming backend tasks (mq, compaction).

So the load should consist of 
- from leo_gateway(s) generated by basho_bench
- from leo_storage(s) generated by (compaction|rebalance)

Please give it another try with the above considerations.

**Comment by windkit** [comment]
@mocchira This is more or less just to demonstrate the performance hit of the watchdog.

I will check the status during compaction anyway.

**Comment by mocchira** [comment]
@windkit 

To make it clear,
The purpose is NOT gaining performance but reduce disk I/O against leveldb in order to prevent badarg from happening on eleveldb.

That being said,
In this test case, 
Decreasing **I/O Write Rate** is more important than throughput itself.
(also it would be great to be able to avoid spikes so we may have to tackle this spike problem with enabled leo_watchdog as another issue)

**Comment by windkit** [comment]
@mocchira Of course it is not going to improve the performance and is not solving the problem.

This is what I observed when I was preparing to run the compaction here.
I just want to show the performance hit to normal operation here.
It is not just a 10% hit but rather a huge one.

I don't like the spiky graph, unpredictable performance is the problem I have with the watchdog.
That's another issue anyway.

I will move on to check whether the watchdog helps to fix the original problem.

**Comment by vstax** [comment]
If you don't mind my opinion about this - I don't know much about LeoFS internals, but I have some knowledge about storage systems and some typical problems - I'd like to explain why the metric for this watchdog is problematic in general (so much that it might be not possible to make it work correctly without carefully tweaking for each system).

"Device utilization" reported by iostat is, basically, amount of milliseconds when there was something in IO queue for the device divided by 1000 (total milliseconds). Good explanation here: http://stackoverflow.com/a/8512978; also, this is from "iostat" man page:
```
%util
       Percentage  of  elapsed time during which I/O requests were issued to the device (bandwidth utiliza‐
       tion for the device). Device saturation occurs when this value is close to 100% for devices  serving
       requests  serially.   But  for  devices serving requests in parallel, such as RAID arrays and modern
       SSDs, this number does not reflect their performance limits.
```
Note the last sentence; but the problem with this metric is even bigger. This might be a good metric of how device is saturated from sysadmin perspective, but incorrect metric for application itself because it has nothing to do with what application actually needs - answer to the question "is it fast enough or not?"

Consider a SSD which is under nearly 100% load, according to iostat; it's easy to reproduce. This is on loaded PostgreSQL database (I have merged output a bit for better reading):
```
# iostat -xkz /dev/sda 1
Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda              18,03     7,50  185,03  173,29  4640,12  1490,19    34,22     0,00    0,01    1,29    0,60   0,10   3,65
sda               5,00     0,00  339,00  130,00  9768,00   520,00    43,87     0,15    0,32    0,42    0,05   0,12   5,50
sda               0,00     0,00    4,00  172,00   112,00  1160,00    14,45     0,02    0,11    0,00    0,11   0,11   1,90
sda               1,00    56,00   18,00  185,00   788,00  1325,00    20,82     0,02    0,10    0,22    0,09   0,10   2,00
```
Adding some fake load:
```
# dd if=/dev/sda of=/dev/zero bs=1M iflag=direct &
# iostat -xkz /dev/sda 1
Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0,00     0,00 1034,00  178,00 529408,00   996,50   875,25     1,81    1,49    1,45    1,76   0,82  99,40
sda               0,00     0,00 1022,00  182,00 518224,00   931,00   862,38     1,82    1,51    1,48    1,70   0,82  99,20
sda               0,00    31,00 1050,00  199,00 533816,00  1166,00   856,66     1,90    1,52    1,43    1,98   0,79  99,10
```
Disk utilization is over the top, but the actual metrics related to performance (r_await and w_await) are hardly any worse; e.g. the database does the same amount of writes (w/s) and doesn't suffer a bit! Moreover, r_await and w_await are average (a) wait times; times from that "dd" process are
included in r_await and possibly make it worse, while wait time for database itself might be not affected - or maybe they are, there is no way of knowing from "iostat". Still, %util doesn't reflect the performance from the point of application here at all and using it for any logic would produce bad results.

PostgreSQL has no logic to lower the load when IO is slow, but some distributed services do. For example, Zookeeper checks time required for fsync WAL log after some operation. If it takes more time than some value, the operation is aborted (only on this node - if over the half of nodes accept the operation, all is fine, otherwise client gets abort). Actually, recent versions have a config value (fsync.warningthresholdms). Default values are something like - fsync that takes over 1 second there is a warning, if it takes over 20 seconds or so (not sure; I've seen this when it was 24) - operation is aborted.

Another example: HDFS namenodes request operations from journalnodes, which require fsync for their operation to complete. If fsync operation takes too long, the operation of aborted. The end result depends on whether enough operations have succeeded. If writing to all journals have failed because of timeout, the write operation is impossible and filesystem crashes / restarts. This is rather extreme case, but if just some fsync operations are taking too much time, there are warnings as well: http://stackoverflow.com/questions/35025934/hdfs-journal-node-giving-alerts-fsync-latency-concerning

The problem with disk utilization from iostat - even when not taking SSDs or high-performance RAIDs into account - is that it's too indirect. For application logic to detect the problem and do something about it, more direct logic is needed. E.g. for the case of "if this operation takes too much time, there will be problems" the direct logic is "monitor time of that operation - read time for reads, fsync time for synced writes -> react if it exceeds acceptable amount". Indirect logic like %util is, well, too indirect, it can't be used as a feedback reliably.

Whether it's possible to use such direct reading in LeoFS or not, is, of course, another question... I think (I might be wrong here, this is just a speculation) that it might be possible to find "that one" critical operation - or a few of them - and monitor its time, and use that for "disk utilization watchdog" instead. E.g. if the problem is "leveldb compaction creates STW situation and causes restart if it takes too long", then the correct metric of choice might be fsync() from that compaction process. It might be something else, of course. Just, well, replacing indirect logic "IO load is too high -> operations are too slow -> monitor device utilization -> if it's too high, decide that everything is too slow and do something to lower the load" with direct "operations are too slow -> monitor operations times, if it exceeds certain amount, do something to lower the load" might produce much nicer results (and won't require too much tweaking of watchdog parameters depending on performance of IO subsystem for each case, checking the direct metrics takes care of that).

**Comment by mocchira** [comment]
@windkit 

> I don't like the spiky graph, unpredictable performance is the problem I have with the watchdog.
That's another issue anyway.

Yes I also don't. let's move this topic on the another issue.

@vstax thanks always for your comment in detail.

> replacing indirect logic "IO load is too high -> operations are too slow -> monitor device utilization -> if it's too high, decide that everything is too slow and do something to lower the load" with direct "operations are too slow -> monitor operations times, if it exceeds certain amount, do something to lower the load" might produce much nicer results (and won't require too much tweaking of watchdog parameters depending on performance of IO subsystem for each case, checking the direct metrics takes care of that).

Good point.
Now we already have a watchdog measuring I/O response time and reduce the load according to its result however it's only measuring I/O originally caused by leo_gateway (online tasks) not measuring I/O caused by leo_storage (offline tasks) like consuming queues and proceeding compaction so the current leo_watchdog couldn't be responsive in your case (according to the log you would share on #644, consuming queues were seen around restarts).

So measuring I/O response time on offline tasks might be more promising solution.
We are going to discuss again while taking your comment into account.

Thanks :)

**Comment by windkit** [comment]
I think the problem here is not about what threshold should be used but we need to review the watchdog design for monitoring disk load

**Comment by mocchira** [comment]
@windkit maybe.

For this reason,
As we need a more general topic to discuss about leo_watchdog improvements
- Responsiveness
  - In some cases, triggering the leo_watchdog in realtime might be desirable.
- Less Spiky
  - Ideally SHOULD decrease ONLY throughput, keep other metrics stable as is.  
- Less Configuration
  - Won't require too much tweaking of watchdog parameters for Sys Admin.

I will file this as another issue later with milestone 1.4.
